Cloud Custodian Joint Review markdown (#786)
cncf/tag-security: assessments/projects/custodian/joint-review.md

* [Intended use](#intended-use)
* [Project design](#project-design)
  * [Functions and features](#functions-and-features)
    * [Security functions and features](#security-functions-and-features)
* [Configuration and set-up](#configuration-and-set-up)
* [Project compliance](#project-compliance)
* [Security analysis](#security-analysis)
* [Secure development practices](#secure-development-practices)
* [Security issue resolution](#security-issue-resolution)
  * [Closed security issues and vulnerabilities](#closed-security-issues-and-vulnerabilities)
* [Roadmap](#roadmap)
* [Appendix](#appendix)

## Metadata

| Project Home                    | https://github.com/cloud-custodian                                                                               |
|---------------------------------|------------------------------------------------------------------------------------------------------------------|
| Security Provider               | Yes, however many users implement Cloud Custodian for non security purposes, i.e., cost optimization, governance |
| Languages                       | Python, YAML |
| Incubation PR                   | https://github.com/cncf/toc/pull/480 |

> | SIG-Security Assessment Request | https://github.com/cncf/sig-security/issues/307 |

Reviewer: ```suggestion
| TAG-Security Assessment Request | https://github.com/cncf/sig-security/issues/307 |
```

> | Self-Assessment Document        | https://docs.google.com/document/d/1s88ifDtFJzGX1O_ve6HM6_vbTkpsWDKWRDuxt9RHdcs/ |

Reviewer: I believe this document has to be converted as well and checked into the PR

Reviewer 2: Yep - it does, they should both go in the folder.

Reviewer 3: Besides the self assessment doc, is there anything else that would need to be archived for posterity? (i.e. minutes/checkins, etc.)?

Author: do we need the project's approval for that?

Reviewer 2: The self-assessment should be owned by the project and submitted as a PR to our repo for independent attestation by the security group or minimally have the markdown linked in these docs

| SBOM                            | [poetry configuration in pyproject.toml](https://github.com/cloud-custodian/cloud-custodian/blob/master/pyproject.toml), [poetry lock file](https://github.com/cloud-custodian/cloud-custodian/blob/master/poetry.lock), [pip requirements file](https://github.com/cloud-custodian/cloud-custodian/blob/master/requirements.txt), [release process](https://cloudcustodian.io/docs/developer/packaging.html) |

> 

Reviewer: Can we add list of reviewers / project participants for the joint review (we should fix the template to include this)

Reviewer 3: I'm thinking a copy of our initial conflict-of-interest disclosures (from the GH issue/ticket) could also be included verbatim in some appendix at the end.

Reviewer: >I'm thinking a copy of our initial conflict-of-interest disclosures (from the GH issue/ticket) could also be included verbatim in some appendix at the end.

it probably is not necessary , it just needs a chair sign off on the issue. it should all be saved in Github history

Reviewer 3: Sounds good.

Reviewer 2: I have an open PR #787  with the correct table to add in the reviewers.

Author: I think this is resolved now...please review

# Cloud Custodian Joint-review

This joint-review relied heavily on the [self-assessment](https://docs.google.com/document/d/1s88ifDtFJzGX1O_ve6HM6_vbTkpsWDKWRDuxt9RHdcs)
provided by the Cloud Custodian (c7n) project. The Security TAG volunteers collaborated with the project volunteers to review the self-assessment, 
and discussed the operational scenarios pertinent to a typical c7n user.

The general opinion of the volunteer review team is that the c7n project clearly demonstrates a commitment to security, and a willingness to improve security for the benefit of the user community.  That said, the project volunteers could use additional security volunteers to assist in ongoing reviews, security control 
enhancements, and when needed to debug and fix security flaws.  As a utility tool, the review team feels the primary threat to the project is compromise of the code

> itself, and packaging and distribution of the tool or its dependencies - rather than an attack on individual c7n deployments. A secondary concern would the disabling

Author: fixed

| Threat Model (2019) | https://docs.google.com/document/d/1S9zQZaT6G1TA3IAx6YNL0f7G938xaFZ-bziszhuxMZg |

## Overview

Cloud Custodian is a YAML DSL-based stateless rules engine for cloud auditing, management, and governance.

Cloud Custodian’s policy-as-code:

* Enables deployment of detection controls that enforce or report compliance to organizational policies.
* Supports real-time detection, reporting/notification, and remediation via an API audit trail integration and serverless runtime support.
* Provides consistent outputs and telemetry (blob, logs, trace, metrics) across policies with outputs to provider native services.
* Requires minimal installation requirements, and can be used as a CLI query/investigative or operations tool in a compliance-as-code environment.

### Background

Cloud Custodian is a YAML DSL-based stateless rules engine operated via a CLI, easily triggered from a cloud-resident cron task or containerized workload. 
It supports the practice of policy-as-code — policy being expressed as rules that the CLI uses to check cloud configurations.  
The code artifacts serve as documentation of the configuration policies in effect in a given environment.

> NOTE: Humans must map written compliance or security documented policies to specific Custodian YAML first.

Reviewer 3: "Cloud Custodian" or "c7n".

|                      |                  |          |                          |               |                                |
|                      |                  |          |                          |               |                                |
+----------------------+                  +          +                          +---------------+                                +
| python3.8-venv       |                  |          |                          |               |                                |
|                      |                  |          |                          |               |                                |
|                      |                  |          |                          |               |                                |
|                      |                  |          |                          |               |                                |
+----------------------+------------------+----------+--------------------------+---------------+--------------------------------+
```
 
## Hands-on Review

> Although some of the reviewes have previously used c7n in AWS, both as cli and in Lambda, Cloud Custodian did not receive a join hands-on review from TAG-Security.

Reviewer 4: Typo: `reviewes`

Author: fixed

 
* The project should determine if operators can run separate instances of c7n for detection vs. remediation, and document how to best deploy under segregation of duties principles. This may minimize "blast radius" in the event that a detective c7n environment is compromised.  Presumably the remediation environment, and policies, could be more tightly controlled, and more meticulously scrutinized and tested and monitored.
 
* Finally, the review team suggests that Kubernetes should receive more attention in the project roadmap, as well as support for other CNCF projects, as well as non-CNCF open source, cloud native frameworks (e.g. Istio).

_Recommended CNCF Requests_ 

The review team recommends that CNCF continue to offer fuzzing resources, and hopes the project avails itself of these.  We also recommend a broader set of projects review and deploy c7n to document more open source use cases.  Finally, we hope that CNCF will fund a formal code audit and lab setup assessment which can subject c7n to real world attacks in a Kubernetes environment, once the project is ready for this with full Kubernetes support.
 
## Appendix

### Case Studies
 

> TODO: 2-3 scenarios of real-world use cases.

Reviewer 4: This will be resolved before we merge this correct ?

Author: will ping Jorge and @kapilt

Reviewer 2: @sunstonesecure-robert @castrojo This is still outstanding.  Please add 2-3 real world use cases before we can merge in, otherwise everything is good to go.

This joint-review relied heavily on the
[self-assessment](https://docs.google.com/document/d/1s88ifDtFJzGX1O_ve6HM6_vbTkpsWDKWRDuxt9RHdcs)
provided by the Cloud Custodian (c7n) project. The Security TAG volunteers
collaborated with the project volunteers to review the self-assessment, and
discussed the operational scenarios pertinent to a typical c7n user.

The general opinion of the volunteer review team is that the c7n project
clearly demonstrates a commitment to security, and a willingness to improve
security for the benefit of the user community.  That said, the project
volunteers could use additional security volunteers to assist in ongoing
reviews, security control enhancements, and when needed to debug and fix
security flaws.  As a utility tool, the review team feels the primary threat to
the project is compromise of the code itself, and packaging and distribution of
the tool or its dependencies - rather than an attack on individual c7n
deployments. A secondary concern would the disabling or DoS of a deployment by
an internal or external actor - thus giving the user a false sense of security
when c7n is disabled. A tertiary concern is the provisioning of too permissive

> a policy within the cloud environment for c7n; this would make it an attrative

Reviewer 2: ```suggestion
a policy within the cloud environment for c7n; this would make it an attractive
```

| Languages                       | Python, YAML |
| Incubation PR                   | https://github.com/cncf/toc/pull/480 |
| TAG-Security Assessment Request | https://github.com/cncf/tag-security/issues/307 |
| Self-Assessment Document        | https://docs.google.com/document/d/1s88ifDtFJzGX1O_ve6HM6_vbTkpsWDKWRDuxt9RHdcs/ |
| SBOM                            | [poetry configuration in pyproject.toml](https://github.com/cloud-custodian/cloud-custodian/blob/master/pyproject.toml), [poetry lock file](https://github.com/cloud-custodian/cloud-custodian/blob/master/poetry.lock), [pip requirements file](https://github.com/cloud-custodian/cloud-custodian/blob/master/requirements.txt), [release process](https://cloudcustodian.io/docs/developer/packaging.html) |

## Reviewers and Participants

* Security Review Lead:
* Security Co-Lead:
* Security Co-Lead:
* Security Reviewer:
* Security Reviewer:
* Project Lead:
* Project Member:
* Project Member:

### Security links

| Item | Ref or URL |
| -- | -- |
| Security Mailing List  | security@cloudcustodian.io |
| Threat Model (2019) | https://docs.google.com/document/d/1S9zQZaT6G1TA3IAx6YNL0f7G938xaFZ-bziszhuxMZg |

## Overview

> Cloud Custodian is a YAML DSL-based stateless rules engine for cloud auditing,

Reviewer 2: Please expand DSL here then abbreviate in parens.

Author: fixed

the CLI uses to check cloud configurations.  The code artifacts serve as
documentation of the configuration policies in effect in a given environment.

NOTE: Humans must map written compliance or security documented policies to
specific Cloud Custodian YAML first.

At the moment, Cloud Custodian supports the most common public cloud providers:
AWS, Azure, and GCP. Executed against these environments, with the
corresponding YAML rules, Cloud Custodian can check cloud settings via cloud
APIs for encryption and access requirements, tagging policies, cost management,
removal of abandoned resources, and automated management of resources during
off-hours.

Cloud Custodian currently has [prototyped Kubernetes cluster support as
documented here](https://github.com/cloud-custodian/cloud-custodian/pull/2760).

### Project Goals

Cloud Custodian integrates with the cloud APIs of each provider to provide
checks for enforcement of resource provisioning and configuration policy goals.
It can be run as a simple cron job on a server to execute against existing
resources.  Cloud Custodian is essentially a rules engine for managing public

> cloud accounts and resources. It allows users to define custom yaml policies as

Reviewer 2: nit: YAML v. yaml should be consistent throughout the doc.

Author: the official YAML docs use all caps YAML, so made consistent

  engineering background can still use it. The same Custodian outputs the
  Enterprise Operator can use to produce a centralized survey of an
  organization’s resources and resource activity, a Quota Operator can use to
  determine resource boundaries from the same reports and metrics. When a
  resource exceeds its quota, Cloud Custodian can be configured to notify a
  Quota Operator of it.
* Third-Party Security Product/System: Cloud Custodian’s YAML configuration
  file enables resource tagging based on filters, and depending on the policy,
  these tags can be configured to trigger specific actions.
* Platform Implementer: Since Cloud Custodian enables event-based real-time
  policy enforcement, a Platform Implementer can use the CLI to serve the
  intersecting needs of the Enterprise Operator, Quota Operator, Security and
  Compliance Administrators, and Developers. Event-based and real-time Cloud
  Custodian policies can output reports and metrics for monitoring, respond to
  policy changes and tag affected resources, enforce compliance, and prevent
  development from unintentionally harming an organization’s cloud.

> ### Target Capabilities and How Custodian Impacts.

Reviewer 2: ```suggestion
### Target Capabilities and How Custodian Impacts
```

  components running and minimal inbound access.
* Custodian requires cloud credentials (role, service account/principal) with
  elevated permissions for introspection and remediation. Access to those
  credentials should be restricted to the Custodian compute execution
  environment.
* Custodian cloud credentials should be limited in scope to what is necessary
  for custodian policy execution, i.e. minimized access.
* As the policy files are considered trusted inputs -- and per IaaC best
  practices -- an audit trail is implemented to validate the assurance of the
  supply chain through rigorous version and change control mechanisms.

At a project level, to mitigate primary threats to the project and code, the
project uses GitHub’s security alerts (dependabot) scanning for CVEs on the
dependency graph. Additionally, PRs are reviewed by a growing number of
committers to prevent malicious code to be accepted into the code base.

Beyond these, there are few security-specific features, as the project assumes
the host cloud and corporate environment itself is hardened and trusted.  It
remains to see how useful these assumptions would be in a Zero Trust

> Architecture (ZTA).

Reviewer 2: Is there a way we can balance this?  Earlier on we discuss how c7n was designed with security in mind, but here we don't list ANY of the security design elements that showcase that.  Instead we make it sound like it is insecure and relies entirely on the environment to provide security of the project.
Ideally this would list pivot or leverage point that the c7n team should pay attention to when reviewing PRs as they could negatively (or positively) impact how secure c7n is.

Author: I think this is a real tension in the project vs the TAG goals - c7n consistently and transparently (to its credit) bills itself as a security tool that assumes (and requires) that the user has deep, expert level security knowledge to use the tool correctly.  it's not for novices.  the TAG goal is to make it easy/easier for novice users.  so these are actually out of balance.

Reviewer 2: Your explanation makes more sense to me.  Is there a way we can call that out more, this is a security tool that does requires _increased_ diligence from the user to understand the tools configuration and capabilities throughly?

| Malicious Internal User | An individual, such as an administrator or developer, who leverages their privileged access to Cloud Custodian to create a disruption. |
| Internal Attacker       | An attacker who transited one or more trust boundaries and has access to the cloud provider environment in which Cloud Custodian is hosted in. |
| External Attacker       | An attacker who is external to the cloud provider environment in which Cloud Custodian is hosted in and is unauthenticated. |
| Administrator           | An actual administrator of an account in a given cloud provider environment, tasked with operating and maintaining Cloud Custodian as a whole.     |
| Policy Developer        | A developer who is deploying changes to Cloud Custodian policies, either directly or via source code repositories. |
| Project Developer       | A developer who is deploying changes to Cloud Custodian itself, either directly via custodian’s source code repositories, or by injecting malicious dependency code into the build/packaging/distribution.                                 |
| End User                | An external user of Cloud Custodian. |

### MITRE ATT&CK Review

> Please see: https://attack.mitre.org/matrices/enterprise/cloud/

Reviewer 2: ```suggestion
Please refer to the [Mitre ATT&CK matrix for Enterprise Cloud](https://attack.mitre.org/matrices/enterprise/cloud/)
```

the environment and have reasonable familiarity with their organization’s
deployment of Cloud Custodian. As such, it is expected these attackers will be
intimately aware of the Cloud Custodian configuration, policies, deployment,
serverless usage, and reporting integration. It is also assumed they may retain
ownership permissions to the repository where the configurations for the Cloud
Custodian deployment is kept.

Since c7n does not run as a daemon exposed to the internet (or any network), a
comprehensive analysis of an external attack is not covered within this
document. By definition, a successful external attacker who gains access to the
c7n execution environment will likely take advantage of the same attack paths
as an insider.

### Attack Risks and Effects

Cloud Custodian’s serverless capabilities enable broad access to affect both
the availability and confidentiality of the environment it is deployed within.
Explicit control and review over Cloud Custodian yaml is an absolute necessity
to reduce the success of attack by a malicious insider or opportunistic
external intruder.

As a result of the elevated permissions Cloud Custodian requires in order to

> execute within a given cloud environment, Organizations which fail to

Reviewer 2: ```suggestion
execute within a given cloud environment, organizations which fail to
```

self-denial-of-service.

Both internal and external scenarios vary for different cloud providers. Many
of the conceptual components of a threat scenario are shared between the
clouds, but their manifestations differ based on their specific
implementations.

While external attackers are also possible - especially if custodian is running
in a persistent node (e.g. EC2 instance running cron), or even if it is running
in a docker container that pulls the latest code from the project which was
compromised - a comprehensive analysis of an external attack is not covered
within this document. A successful external attack will gain access to the
environment and take advantage of the same attack paths as an insider, with a
notable exception that they may not also have access to the policy code
repository. That being said, if an external attacker were to successfully have
improper configurations and policies merged into the source code of the
organization’s Cloud Custodian deployment repository, they may grant themselves
broader access into an environment and therefore fully replicate insider attack
possibilities.

### Example Attack Scenarios

> * Denial of Service, Tampering: Use lambda:PutFunctionConcurrency  to set a

Reviewer 2: ```suggestion
* Denial of Service, Tampering: Use `lambda:PutFunctionConcurrency` to set a
```

>   Lambda Function's concurrency setting to 0 which would disable event based

Reviewer 2: ```suggestion
  Lambda Function's concurrency setting to `0` which would disable event based
```

  compliance checking.
* Information Disclosure: Redirecting Cloud Custodian log output

>   (logs:CreateLogStream, logs:PutLogEvents, logs:CreateLogGroup) to a

Reviewer 2: ```suggestion
  (`logs:CreateLogStream`, `logs:PutLogEvents`, `logs:CreateLogGroup`) to a
```

  misconfigured or non-compliant S3 bucket (CLI: `custodian run --output-dir
  s3://<my-bucket>/<my-prefix> <policyfile>.yml`), revealing to an attacker
  detailed metadata about the cloud environment as well as positive Cloud
  Custodian findings on how it is non-compliant (i.e. vulnerable).
* Spoofing, Elevation of Privileges: Use lambda:CreateFunction and

>   lambda:InvokeFunction, thereby allowing for arbitrary code to be introduced

Reviewer 2: ```suggestion
  `lambda:InvokeFunction`, thereby allowing for arbitrary code to be introduced
```

  and executed when triggered.

> * Alternatively, use lambda:UpdateFunctionCode to update code for an existing

Reviewer 2: ```suggestion
* Alternatively, use `lambda:UpdateFunctionCode` to update code for an existing
```

>   Lambda Function which would bypass the need for iam:PassRole. Then use

Reviewer 2: ```suggestion
  Lambda Function which would bypass the need for `iam:PassRole`. Then use
```

>   lambda:InvokeFunction to invoke whatever code that you want using the IAM role

Reviewer 2: ```suggestion
  `lambda:InvokeFunction` to invoke whatever code that you want using the IAM role
```

>   that the Lambda Function is already attached to.  The iam:PassRole permissions

Reviewer 2: ```suggestion
  that the Lambda Function is already attached to.  The `iam:PassRole` permissions
```

  are only checked at function creation, not when the function is updated.

> * Attacker uses a Cloud Custodian Lambda function to leverage (iam:PassRole) to

Reviewer 2: ```suggestion
* Attacker uses a Cloud Custodian Lambda function to leverage (`iam:PassRole`) to
```

  CloudTrail Mode (AWS lambdas are triggered by cloudtrail events to terminate
  things). Example one policy that terminates instances with encryption turned
  on and another that kills instances without encryption.

### Security Degradation

The project itself is a key target. With a small community, and even smaller
approver and project review team, the likelihood that malicious policy content
or code or dependencies will be released is non-zero.  This would significantly
compromise users of Custodian who regularly update from the project git
repository or use their python package distribution.

The policy content maintained by the organization is a secondary target.
Except at very large organizations, it is likely 1 or 2 users would maintain
the entire policy repository with little to no oversight. It is unrealistic to
expect smaller organizations to effectively review every policy change.

The cloud permissions granted to Custodian are broad. Neither a secure default
policy, nor policy generation or checker/linter tool, nor even adequate
documentation exists to guide the new user on how to effectively lock down

> their policy to the absolute minimum.  Even then, the mixing of broad read and

Reviewer 2: Recommend adding another Recommendation about the project that a secure default policy, linter/checker, and additional documentation be created to guide new users on locking down their policies.

Author: @TheFoxAtWork added to the CNCF ask specifically. please review

write permissions in one policy is itself a risk.  Attackers are likely to use
custodian once inside, via other attacks, to move laterally and cover their
tracks.

Probably the largest impact of an attack would be to allow attackers to

> effectively disable custodian and prevent detection or remediation of their

Reviewer 2: nit : Custodian is lower case here in stead of upper. Should be consistent

  possibly, efficacy/safety check(s). This may prevent casual tampering of yaml
  policies once deployed in an organization.
* The project should determine if operators can run separate instances of c7n
  for detection vs. remediation, and document how to best deploy under
  segregation of duties principles. This may minimize "blast radius" in the
  event that a detective c7n environment is compromised.  Presumably the
  remediation environment, and policies, could be more tightly controlled, and
  more meticulously scrutinized and tested and monitored.
* Finally, the review team suggests that Kubernetes should receive more
  attention in the project roadmap, as well as support for other CNCF projects,
  as well as non-CNCF open source, cloud native frameworks (e.g. Istio).

### Recommended CNCF Requests

The review team recommends that CNCF continue to offer fuzzing resources, and
hopes the project avails itself of these.  We also recommend a broader set of
projects review and deploy c7n to document more open source use cases.
Finally, we hope that CNCF will fund a formal code audit and lab setup
assessment which can subject c7n to real world attacks in a Kubernetes

> environment, once the project is ready for this with full Kubernetes support.

Reviewer 2: Maybe worth mentioning here that CNCF assist with documentation resources to better enable access to security hardening and policy information regarding the project.

Author: are there CNCF resources available for docs activities? or more just engaging with the volunteer community to solicit docs review? or ... ?

Reviewer 2: https://github.com/cncf/servicedesk

- Foundation
- Design and Aesthetics
- Program Management
- Legal Services
- Tools
- **Technical Documentation**
- Continuous Integration
- Certification and Training Services
- End User Community
- Internationalization
- Marketing (event management, marketing services and programs, marketing communications)
- Marketing announcements for projects

<!-- cSpell:ignore attck addgetnetgrent JPMC -->
# Cloud Custodian Joint-review

This joint-review relied heavily on the
[self-assessment](https://docs.google.com/document/d/1s88ifDtFJzGX1O_ve6HM6_vbTkpsWDKWRDuxt9RHdcs)
provided by the Cloud Custodian (c7n) project. The Security TAG volunteers
collaborated with the project volunteers to review the self-assessment, and
discussed the operational scenarios pertinent to a typical c7n user.

The general opinion of the volunteer review team is that the c7n project
clearly demonstrates a commitment to security, and a willingness to improve
security for the benefit of the user community.  That said, the project
volunteers could use additional security volunteers to assist in ongoing
reviews, security control enhancements, and when needed to debug and fix
security flaws.  As a utility tool, the review team feels the primary threat to
the project is compromise of the code itself, and packaging and distribution of

> the tool or its dependencies - rather than an attack on individual c7n

Reviewer 4: I'm curious what led to this conclusion since we haven't done a code review or tried out a sample deployment.

|                      |                  |          |                          |               |                                |
|                      |                  |          |                          |               |                                |
|                      |                  |          |                          |               |                                |
+----------------------+                  +          +                          +---------------+                                +
| python3.8-venv       |                  |          |                          |               |                                |
|                      |                  |          |                          |               |                                |
|                      |                  |          |                          |               |                                |
|                      |                  |          |                          |               |                                |
+----------------------+------------------+----------+--------------------------+---------------+--------------------------------+

> ```

Reviewer 4: Do we need the `HIGH` vulnerabilities to be resolved or at least require the c7n team to open issues for them and link them here ?

Reviewer 5: These are operating system library level considerations not per se the project's responsibility to fix, but the distribution vendor, ie a libc high issue is something i would expect the os vendor to triage and fix as needed. This is also specific to a distribution channel, and immaterial for those that install the project's software directly into their own OS / container base. For direct application dependency graphs concerns we already address those pro actively (dependabot alerts etc).

Reviewer 6: these were from the Github action docker scan - if the project is releasing/maintaining a docker image I think the user expectation is they should be fixing high security risks in the image as good hygiene.

maybe best to just do source releases? or clearly document that the image is not secure?

Reviewer 5: This feels like a misunderstanding of operating system docker images, and what severity levels correspond to a need to immediate triage. If we take any production docker image out there using an operating system, they all have these sorts of cve nits wrt to `high` severity scoring. Take anything from redis to postgresql, (trivy image redis:latest etc), that doesn't make them insecure or high risk (trivy output for both here https://gist.github.com/cbdd697ce7f8ad0a40896c0a081c1a7c). The most relevant cves are those deemed critical, but all of operating system cves are triaged and dealt with by the security team of the respective distros which in many cases are professional security teams focused on exactly this topic. Moreover unlike postgresql or redis, we're talking about software in custodian that doesn't have an open port, its literally a cli, not a daemon or service running in a container.

Reviewer 4: My original comment was related to direct dependancies of c7n. @kapilt I think you mentioned you have a process to handle those iiuc.

Reviewer 5: for custodian's dependency graph, we rely on dependabot alerts https://docs.github.com/en/code-security/supply-chain-security/managing-vulnerabilities-in-your-projects-dependencies/about-alerts-for-vulnerable-dependencies against our poetry.lock files which contain the version/hash dependency pins for the project.  that alert is a prompt for us to update our dependency graph. our dep graph update process is automated and documented https://cloudcustodian.io/docs/developer/packaging.html and done regularly independent of vulnerabilities as a prelude to our release process.

Reviewer 5: also to be specific there are zero cves extant on our application dependency graph at the moment.

## Hands-on Review

Although some of the reviews have previously used c7n in AWS, both as cli and

> in Lambda, Cloud Custodian did not receive a join hands-on review from

Reviewer 4: Nit: `joint`

> in Lambda, Cloud Custodian did not receive a join hands-on review from

Reviewer 2: ```suggestion
in Lambda, Cloud Custodian did not receive a joint hands-on review from
```
